File: alicoding/34.java
package solution34;

class Solution {
    private int idx = 0;
    private int[] countArr;
    public int solution(int[][] m) {
        this.countArr = new int[m.length*m[0].length*2];
        this.subSolu(m, 0, 0, m[0][0]);
        int min = this.countArr[0];
        for (int i=1;i<this.countArr.length;i++) {
            if (min > this.countArr[i] && this.countArr[i] > 0) {
                min = this.countArr[i];
            }
        }
        return min;
    }

    public void subSolu(int[][] m, int i,int j,int count ){ 
        boolean isDown = false, isRight = false ;
        if (i>=m.length-1){
            isDown = true;
        }
        if (j>=m[0].length-1){
            isRight = true;
        }
        if (isDown && isRight) {
            this.countArr[this.idx] = count;
            this.idx++;
        }
        if (!isDown ) {
		    this.subSolu(m, i+1, j, count + m[i+1][j]);
        }
        if (!isRight) {
		    this.subSolu(m, i, j+1, count + m[i][j+1]);
        }
    }
}

System.out.println("hei");

File: alicoding/34.go
package main

import (
	"fmt"
	"sort"
)

var (
	countArr = []int{}
)

func main() {
	in := [][]int{
		[]int{4, 1, 5, 3},
		[]int{3, 2, 7, 7},
		[]int{6, 5, 2, 8},
		[]int{8, 9, 4, 5},
	}
	soluSub(in, 0, 0, 4)

	fmt.Println()
	sort.Ints(countArr)

	for i, v := range countArr {
		fmt.Println(i+1, v)
	}

}

func soluSub(arr [][]int, i, j, count int) {
	fmt.Println(i, j, count)
	rightCount, downCount := count, count
	isDown, isRight := false, false
	if i < len(arr)-1 {
	} else {
		isDown = true
	}
	if j < len(arr[0])-1 {
	} else {
		isRight = true
	}
	if isRight && isDown {
		countArr = append(countArr, downCount)
		fmt.Println("Over", rightCount, downCount)
	}

	if !isDown {
		downCount += arr[i+1][j]
		soluSub(arr, i+1, j, downCount)
	}
	if !isRight {
		rightCount += arr[i][j+1]
		soluSub(arr, i, j+1, rightCount)
	}

	return
}
